Expand the rpi_gpio component to support orangepi as well.

mikalstill opened the pull request:
## Description:

Instead of adding a new component for orangepi (which is a raspberry pi clone), let's just extend the existing component to support more than one family of boards.

I have tested this patch on a raspberry pi 3 and an orangepi prime, in both cases with a LED on a GPIO controlled switch. Note that the orangepi GPIO library requires access to /dev/mem, which I have resolved for now by running Home Assistant as root.

I'm holding off on a home-assistant.io pull request to match this until I get a feel for if this approach is acceptable to you.

This code was implemented to support a Home Assistant tutorial running at linux.conf.au 2019 in a couple of weeks, it would be super cool to be able to have this merged by then so that attendees don't need to carry private patches.

## Example entry for `configuration.yaml` (if applicable):
```
rpi_gpio:
  board_family: orange_pi
  board: prime

switch:
 - platform: rpi_gpio
   ports:
     29: LED
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (see PR https://github.com/home-assistant/home-assistant.io/pull/8185)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mikalstill committed:
Expand the rpi_gpio library to support orangepi as well.

Instead of adding a new component for orangepi (which is a
raspberry pi clone), let's just extend the existing component
to support more than one family of boards.

I have tested this patch on a raspberry pi 3 and an orangepi
prime, in both cases with a LED on a GPIO controlled switch.
Note that the orangepi GPIO library requires access to /dev/mem,
which I have resolved for now by running Home Assistant as root.

homeassistant commented:
Hi @mikalstill, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19732).

Once you do that we will be able to review and accept this pull request.

Thanks!

mikalstill commented:
Ummm, I'm happy to sign the CLA, but https://www.home-assistant.io/developers/cla/ implies there should be a link above for me to do that thing, and I can't see it...

M0les committed:
Attempt to fix linter style nags

rohankapoorcom commented:
@mikalstill, it looks like your link is https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant/home-assistant#19732

The message from the home assistant bot had it (the word here was hyperlinked to that url).

mikalstill commented:
I swear that comment wasn't there before. Regardless, I'll sign the CLA now.

homeassistant commented:
Hello @mikalstill,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`7fca000bf913d0af8818c2304ee5c3d19261ab4a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;19732&#x2F;commits&#x2F;7fca000bf913d0af8818c2304ee5c3d19261ab4a) - This commit has something that looks like an email address (miles@m0les.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

mikalstill committed:
Fix trailing whitespace error.

fabaff commented:
> Note that the orangepi GPIO library requires access to /dev/mem, which I have resolved for now by running Home Assistant as root.

To be honest, I not a big fan of suggesting to run HA as root. How was this issue solved for RPi.GPIO?

fabaff commented on homeassistant/components/rpi_gpio.py:
```diff
@@ -5,64 +5,131 @@
+
+def setup(hass, base_config):
+    """Set up the GPIO component."""
+    global GPIO_LIBRARY
```
Don't use `global`, we have `hass.data`.

mikalstill replied:
Ok, I'll go have a look at examples.

mikalstill replied:
Done.

fabaff commented on homeassistant/components/rpi_gpio.py:
```diff
@@ -5,64 +5,131 @@
+        board = ORANGEPI_BOARDS.get(board_name)
+        _LOGGER.info('OrangePi %s board specified', board_name)
+
+        if not board_name:
```
All validation should be done with the Schema.

mikalstill replied:
board_name is only required if you specify an Orange Pi, there is no equivalent concept with Raspberry Pi GPIOs. Is there an existing example of how to conditionally validate that a config value is set dependant on another config value?

mikalstill commented:
> > Note that the orangepi GPIO library requires access to /dev/mem, which I have resolved for now by running Home Assistant as root.
> 
> To be honest, I not a big fan of suggesting to run HA as root. How was this issue solved for RPi.GPIO?

Totally agreed. That said, its not an as-root requirement, its a with-permissions-for-/dev/mem requirement. That can be done with filesystem permissions if required.

My understanding is that Raspberry Pi deals with this by having a kernel driver which exposes the GPIOs in a more friendly manner. I'm not opposed to doing that for OrangePi, but it will take a little time to do.

mikalstill committed:
Remove globals and use hass.data instead.

Per review comments on PR #19732.

deece commented:
The Rpi.GPIO library first tries /dev/gpiomem (which is Broadcom specific, and has perms granted via udev), then falls back to /dev/mem (which will require perms).

We can solve the access issue for /dev/mem with some udev rules.

Having said that, after some discussion around the office at Ozlabs, /dev/mem is quite dangerous, any process that can access that can effectively access the memory of any process on the system, including the kernel, and library devs who access that should be admonished^H^H^H^H taken aside and taught how to abstract access via kernel driver.

In this case, we should have access to /sys/class/gpio, which in provided by the pinctrl subsystem on the Orange Pi. This library access GPIO via that: https://pypi.org/project/OPi.GPIO/  It will need a patch to support the Orange Pi Prime, but that seems pretty straightforward.

deece commented:
It's worth noting that HassOS currently runs HomeAssistant as root, and passes /dev/mem through to the docker container (not that I recommend we take this path). I'll log this as an issue.

mikalstill committed:
Move to a different OrangePi GPIO library

This library doesn't require root or access to /dev/mem because
it uses the sysfs interface for GPIOs instead. It also doesn't
need to know what OrangePi board you're using, so the configuration
is simpler.

mikalstill commented:
Ok, so this latest patch does a few things:

* Moves to a new GPIO library for OrangePi that does not require root (it uses sysfs not /dev/mem)
* Removes the requirement to configure the OrangePi board, as it turns out they all seem to have the same GPIO pinout

I think that resolves all of fabaff's concerns?

mikalstill committed:
Move to SUNXI pin naming for OrangePI.

Its much less confusing, given there are two competing numbering
schemes.

mikalstill commented:
Hey @pvizeli, it was suggested I ping you about two things around this PR:

- the CLA-Error label is now incorrect, but I don't know how to remove it. How do we do that thing?

- also, the pin numbering scheme for OrangePi that makes the most sense is something called SUNXI. However, its not integers. How would we feel about relaxing the schema for pin numbers in components dependant on this one such as rpi_switch? I'd prefer to do that than to fork out a separate OrangePi switch component if possible.

Thanks!

mikalstill committed:
Update requirements with new OrangePi GPIO library

MartinHjelmare commented on homeassistant/components/rpi_gpio.py:
```diff
@@ -5,64 +5,100 @@
+    """Set up the GPIO component."""
+    hass.data[DOMAIN] = {}
+
+    config = base_config.get(DOMAIN)
```
Use `dict[key]` for required config keys and keys with default config schema values.

MartinHjelmare commented on homeassistant/components/rpi_gpio.py:
```diff
@@ -5,64 +5,100 @@
+    family_name = config.get(CONF_BOARD_FAMILY, DEFAULT_FAMILY)
+    lib_name = FAMILY_LIBRARIES.get(family_name)
+    _LOGGER.info('Configured to use board family %s', family_name)
+    _LOGGER.info('Will use %s as GPIO library', lib_name)
```
Merge these logging messages to one.

MartinHjelmare commented on homeassistant/components/rpi_gpio.py:
```diff
@@ -5,64 +5,100 @@
+    if not lib_name:
+        raise UnknownBoardFamily('Unknown board family: %s'
+                                 % config.get(CONF_BOARD_FAMILY))
+    hass.data[DOMAIN][LIBRARY] = importlib.import_module(lib_name)
```
Cache the imported lib in a local variable that we can use below.

MartinHjelmare commented on homeassistant/components/binary_sensor/rpi_gpio.py:
```diff
@@ -49,30 +51,32 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         self._state = None
 
-        rpi_gpio.setup_input(self._port, self._pull_mode)
+        rpi_gpio.setup_input(self._hass, self._port, self._pull_mode)
```
Move this to entity coroutine method `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/binary_sensor/rpi_gpio.py:
```diff
@@ -49,30 +51,32 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Represent a binary sensor that uses Raspberry Pi GPIO."""
 
-    def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
+    def __init__(self, hass, name, port, pull_mode, bouncetime, invert_logic):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/binary_sensor/rpi_gpio.py:
```diff
@@ -49,30 +51,32 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
             self.schedule_update_ha_state()
 
-        rpi_gpio.edge_detect(self._port, read_gpio, self._bouncetime)
+        rpi_gpio.edge_detect(self._hass, self._port, read_gpio,
```
Register callbacks in `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/cover/rpi_gpio.py:
```diff
@@ -65,17 +65,19 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Representation of a Raspberry GPIO cover."""
 
-    def __init__(self, name, relay_pin, state_pin, state_pull_mode,
+    def __init__(self, hass, name, relay_pin, state_pin, state_pull_mode,
```
See above.

MartinHjelmare commented on homeassistant/components/binary_sensor/rpi_gpio.py:
```diff
@@ -49,30 +51,32 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-    def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
+    def __init__(self, hass, name, port, pull_mode, bouncetime, invert_logic):
         """Initialize the RPi binary sensor."""
+        self._hass = hass
```
Remove this.

MartinHjelmare commented on homeassistant/components/cover/rpi_gpio.py:
```diff
@@ -84,9 +86,11 @@ def __init__(self, name, relay_pin, state_pin, state_pull_mode,
-        rpi_gpio.setup_output(self._relay_pin)
-        rpi_gpio.setup_input(self._state_pin, self._state_pull_mode)
-        rpi_gpio.write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        rpi_gpio.setup_output(self._hass, self._relay_pin)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/switch/rpi_gpio.py:
```diff
@@ -41,21 +43,23 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     """Representation of a  Raspberry Pi GPIO."""
 
-    def __init__(self, name, port, invert_logic):
+    def __init__(self, hass, name, port, invert_logic):
```
See above.

MartinHjelmare commented on homeassistant/components/switch/rpi_gpio.py:
```diff
@@ -41,21 +43,23 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         self._state = False
-        rpi_gpio.setup_output(self._port)
-        rpi_gpio.write_output(self._port, 1 if self._invert_logic else 0)
+        rpi_gpio.setup_output(self._hass, self._port)
```
Move to `async_added_to_hass`.

balloob commented:
This should not be part of the Raspberry Pi integration but instead be a new one. Over time, the integrations will diverge, and we will end up with a bunch of if…else statements. We don't want that.

balloob commented:
This PR seems to have gone stale. Closing it.
